diff --git a/node_modules/@storybook/addon-knobs/dist/registerKnobs.js b/node_modules/@storybook/addon-knobs/dist/registerKnobs.js
index 43b6547..786eacb 100644
--- a/node_modules/@storybook/addon-knobs/dist/registerKnobs.js
+++ b/node_modules/@storybook/addon-knobs/dist/registerKnobs.js
@@ -84,6 +84,7 @@ function resetKnobsAndForceReRender() {
 }
 
 function disconnectCallbacks() {
+  knobStore.reset();
   var channel = _addons["default"].getChannel();
 
   channel.removeListener(_shared.CHANGE, knobChanged);
